### ISSUE TITLE
[freq_table] refactor to be a wrap around dist_matrix_by_dir_sector

### ISSUE THREAD

**Issue by stephenholleran** [brightwind-dev/brightwind#99]
The colours are different from when I merged iss #46 branch into master. I may have messed up something when trying to resolve the conflicts. I think it might be the colour bands changed.

This image is after the merge which has less darker colours.
![image](https://user-images.githubusercontent.com/25622575/59557081-3d870200-8fc8-11e9-8b45-5a8dfa9f84d9.png)

This is before the merge and shows a better wind rose. We should get it back to this.
![image](https://user-images.githubusercontent.com/25622575/59557104-d3229180-8fc8-11e9-8278-b6a3479b5a9c.png)

**Comment by Inder95** [comment]
> The colours are different from when I merged iss #46 branch into master. I may have messed up something when trying to resolve the conflicts. I think it might be the colour bands changed.
> 
> This image is after the merge which has less darker colours.
> ![image](https://user-images.githubusercontent.com/25622575/59557081-3d870200-8fc8-11e9-8b45-5a8dfa9f84d9.png)
> 
> This is before the merge and shows a better wind rose. We should get it back to this.
> ![image](https://user-images.githubusercontent.com/25622575/59557104-d3229180-8fc8-11e9-8278-b6a3479b5a9c.png)

No I think it is fine, you see the bins are also different. We discussed in https://github.com/brightwind-dev/brightwind/issues/46#issuecomment-502113969 of how the commonly used wind rose with the bins in the plot below can be mentioned as an example in the docstring and the default binning option would be equally spaced 6 bins.

**Comment by stephenholleran** [comment]
ok, I think I got a bit lost with that one amongst everything else that was happening in that thread.

If thinking that way then we should do the same with the 'var_bin_array' too.

However in saying that, I think the function should, by default, process for when a wind speed time series is sent.  That is, use bins 0-3, 4-6, 7-9, 10-12, 13-15, 15+.  I think when None is sent that it splits it evenly so we should leave that behaviour. Therefore, we should place the defaults back in for when a wind speed is sent.

**Comment by stephenholleran** [comment]
Tried that and you run into the problem that the '_plot_bins_' and '_plot_labels_' are mutable which after [reading](https://docs.python-guide.org/writing/gotchas/) about it we should avoid.

Thinking about the whole distribution area more and keeping it in line with what we have done with the 2D distributions and @Inder95 your view point that binning by direction sector makes two sets of distributions https://github.com/brightwind-dev/brightwind/issues/46#issuecomment-496227727.

Currently we use `freq_table()` when distributing a variable against itself and direction. This is really a special case for when the wind speed is the variable. Therefore, this should really be just a wrap around a distribution matrix with direction. Similar to `distribution_by_wind_speed()`. Therefore, the '_plot_bins_' and '_plot_labels_' can be set within `freq_table()`.

1. Distributions NOT involving direction

   * For 1 variable distributed against itself we use `distribution()`
   * For special case of wind speed we use `distribution_by_wind_speed()` (and/or `freq_distribution()`
   * For 2 variables to be distributed we use `distribution()`
   * For 2 variables with 1 against itself we will use `distribution_matrix()`
   * For 3 variables we will use `distribution_matrix()`

1. Distributions involving direction

   * For 1 variable distributed against direction we use `distribution_by_dir_sector()`
   * For 1 variable distribution against itself and direction we should use `distribution_matrix_by_dir_sector()`.
   * For special case of wind speed we should use `freq_table()` (and/or `distribution_matrix_by_dir_sector_and_wind_speed()`
   * For 2 variables with 1 against itself and against direction we should use `distribution_matrix_by_dir_sector()`.

List of functions would be:
* `distribution()`
* `distribution_by_wind_speed()` / `freq_distribution()`
* `distribution_matrix()`
* `distribution_by_dir_sector()`
* `distribution_matrix_by_dir_sector()`
* `distribution_matrix_by_dir_sector_and_wind_speed()` / `freq_table()`

Along with `distribution_12x24()`

These names are very long so I propose shortening 'distribution' to 'dist'.

**Comment by stephenholleran** [comment]
I'll rename this issue.

**Comment by Inder95** [comment]
> ok, I think I got a bit lost with that one amongst everything else that was happening in that thread.
> 
> If thinking that way then we should do the same with the 'var_bin_array' too.
> 
> However in saying that, I think the function should, by default, process for when a wind speed time series is sent. That is, use bins 0-3, 4-6, 7-9, 10-12, 13-15, 15+. I think when None is sent that it splits it evenly so we should leave that behaviour. Therefore, we should place the defaults back in for when a wind speed is sent.

I have implemented this change, but as discussed before I still think it makes things more difficult for the user when not dealing with wind speeds.

Now the user has to explicitly specify plot_bins and plot_labels as None or something else
```
bw.freq_table(df.T2m, df.Dir38mS, var_bin_array=[0, 10, 15, 25, 45], plot_bins=None, plot_labels=None)
```
instead of just
```
bw.freq_table(df.T2m, df.Dir38mS, var_bin_array=[0, 10, 15, 25, 45])
```

**Comment by Inder95** [comment]
> Tried that and you run into the problem that the '_plot_bins_' and '_plot_labels_' are mutable which after [reading](https://docs.python-guide.org/writing/gotchas/) about it we should avoid.

Can you show me the code you tried which did not work? Because I tried with arguments followed by None which worked.

**Comment by stephenholleran** [comment]
I did this
![image](https://user-images.githubusercontent.com/25622575/59603831-58f62800-9102-11e9-9ac7-55c8f4348269.png)

It worked but we shouldn't do it as the argument is mutable.

Read my most recent comment. I said we shouldn't do that. We should use freq_table as the wrap around a more generic function allowing us to set the bins within freq_table just like we do for dist_by_wind_speed.

**Comment by Inder95** [comment]
> I did this
> ![image](https://user-images.githubusercontent.com/25622575/59603831-58f62800-9102-11e9-9ac7-55c8f4348269.png)
> 
> It worked but we shouldn't do it as the argument is mutable.
> 
> Read my most recent comment. I said we shouldn't do that. We should use freq_table as the wrap around a more generic function allowing us to set the bins within freq_table just like we do for dist_by_wind_speed.

I have reversed the change, got confused. Will hold off on that generic function for now, will discuss it with you. Working on documentation now.

**Comment by stephenholleran** [comment]
@Inder95 what branch are you working in here? Can't find it.

**Comment by Inder95** [comment]
For understanding how distributions are structured from code perspective, useful for iss #93  too.

![New Doc 2019-06-23 17 17 45](https://user-images.githubusercontent.com/23189856/59979112-39b43a80-95db-11e9-9199-a81002692fcd.jpg)

**Comment by Inder95** [comment]
> @Inder95 what branch are you working in here? Can't find it.

Was working in branch #97

**Comment by stephenholleran** [comment]
> > @Inder95 what branch are you working in here? Can't find it.
> 
> Was working in branch #97

That must be on your local machine?

**Comment by Inder95** [comment]
> > > @Inder95 what branch are you working in here? Can't find it.
> > 
> > 
> > Was working in branch #97
> 
> That must be on your local machine?

I had deleted this branch after merging earlier, restored it back now. But I think I will now work in a new branch, it is a different issue.

**Comment by stephenholleran** [comment]
> > That must be on your local machine?
> 
> I had deleted this branch after merging earlier, restored it back now. But I think I will now work in a new branch, it is a different issue.

Ah, I didn't know you had merged it. Yeah, ditch it and work in a fresh branch.

**Comment by Inder95** [comment]
@stephenholleran wrote a more generic dist_matrix_by_dir_sector(). Could not think of many examples though. Also this would be a good time to change/keep same the default bins for the wind rose plot.

**Comment by stephenholleran** [comment]
- [ ] can you merge master into iss99_dist_matrix_by_dir_sector so I can see dist_matrix or the others? Are the argument names consistent i.e. x_label, y_series? dir_series would obviously replace one of these, the x_series?

- [ ] Tried this `bw.dist_matrix_by_dir_sector(data.Spd80mN, data.Spd80mN, data.Dir78mS, aggregation_method='count')` but got an error.

![image](https://user-images.githubusercontent.com/25622575/61592410-ce09c100-abca-11e9-8022-b6f3cc6aa47b.png)

> Also this would be a good time to change/keep same the default bins for the wind rose plot.

`dist_matrix_by_dir_sector()` should plot the rose in evenly distributed colour bins.
`freq_table()` should plot the wind rose with the colour bins we had before 0,3,6,9,12,15,15+

**Comment by Inder95** [comment]
> * [x]  can you merge master into iss99_dist_matrix_by_dir_sector so I can see dist_matrix or the others and also to plot the timeseries of Shear?

Done

> * [ ]  Are the argument names consistent i.e. x_label, y_series? dir_series would obviously replace one of these, the x_series?

There is no x and y-axis here. When we add plots (if we add any) those will be polar plot, a and y will cause confusion. Also x with y or vice-versa in not intuitive for the user.

> * [x]  Tried this `bw.dist_matrix_by_dir_sector(data.Spd80mN, data.Spd80mN, data.Dir78mS, aggregation_method='count')` but got an error.
 Fixed
> `dist_matrix_by_dir_sector()` should plot the rose in evenly distributed colour bins.
As we discussed plotting this matrix is not easy as we need 3rd axis on a polar plot. One example I found was polar heatmap, still have to find a way a to implement it in python. The other was 3-D map which I feel is far fetched so will drop it.
![image](https://user-images.githubusercontent.com/23189856/61742305-b842fa00-ad8a-11e9-8bb3-85e50c67d902.png)


> `freq_table()` should plot the wind rose with the colour bins we had before 0,3,6,9,12,15,15+. That is, `plot_bins` arg should be default None so the user can overwrite the 0,3,6,9,12,15,15+ which is sent to `dist_matrix_by_dir_sector()`.

Done

**Comment by Inder95** [comment]
@stephenholleran One more question in freq_table we replaced nan by 0. I have done the same for dist_matrix_by_dir_sector but this might not make sense when variable has below 0 values too.

**Comment by stephenholleran** [comment]
> @stephenholleran One more question in freq_table we replaced nan by 0. 

Eh, can't remember. Are you able to elaborate?

> I have done the same for dist_matrix_by_dir_sector but this might not make sense when variable has below 0 values too.

Sounds like it would be a problem if we do shear or temperature which can go negative.

**Comment by stephenholleran** [comment]
> >   `dist_matrix_by_dir_sector()` should plot the rose in evenly distributed colour bins.

>    As we discussed plotting this matrix is not easy as we need 3rd axis on a polar plot. One example I found was polar heatmap, still have to find a way a to implement it in python. The other was 3-D map which I feel is far fetched so will drop it.
>    ![image](https://user-images.githubusercontent.com/23189856/61742305-b842fa00-ad8a-11e9-8bb3-85e50c67d902.png)
> 

Yeah, 3D is probably not needed at this stage.

Interesting plot the heat map. I'm not sure yet how to use it as not able to think deeply about it at this time.

If this doesn't give a plot, how does `freq_table()` get a plot if `freq_table()` is supposed to be a wrap around this function?

**Comment by stephenholleran** [comment]
- [ ] could `var_bin_array` be set to None by default and we guess an array? Didn't we do something like that for dist_matrix?

- [ ] the `variable_bin` name which is the index of the returned table. Could this name be pulled from the name of the series sent?

**Comment by stephenholleran** [comment]
- [ ] `freq_table()` is broken.

![image](https://user-images.githubusercontent.com/25622575/61868817-2219eb80-aed2-11e9-9a1b-dc3f2d140181.png)

**Comment by stephenholleran** [comment]
> > @stephenholleran One more question in freq_table we replaced nan by 0.
> 
> Eh, can't remember. Are you able to elaborate?
> 
> > I have done the same for dist_matrix_by_dir_sector but this might not make sense when variable has below 0 values too.
> 
> Sounds like it would be a problem if we do shear or temperature which can go negative.

Ah, this is to replace empty cells in the returned data and not something to do with redistributing wind speed nans. I have tried putting in '-'. It makes the table a lot more readable as you can quickly see the shape of the data. If we went with this we would just need to take them out when plotting?

![image](https://user-images.githubusercontent.com/25622575/61869467-9a34e100-aed3-11e9-8746-a1e1e28c2907.png)

**Comment by Inder95** [comment]
> * [ ]  `freq_table()` is broken.
> 
> ![image](https://user-images.githubusercontent.com/25622575/61868817-2219eb80-aed2-11e9-9a1b-dc3f2d140181.png)

I am having a look into this, even the previous examples which were working aren't. I think there is a faulty commit somewhere.

**Comment by Inder95** [comment]
> > > @stephenholleran One more question in freq_table we replaced nan by 0.
> > 
> > 
> > Eh, can't remember. Are you able to elaborate?
> > > I have done the same for dist_matrix_by_dir_sector but this might not make sense when variable has below 0 values too.
> > 
> > 
> > Sounds like it would be a problem if we do shear or temperature which can go negative.
> 
> Ah, this is to replace empty cells in the returned data and not something to do with redistributing wind speed nans. I have tried putting in '-'. It makes the table a lot more readable as you can quickly see the shape of the data. If we went with this we would just need to take them out when plotting?
> 
> ![image](https://user-images.githubusercontent.com/25622575/61869467-9a34e100-aed3-11e9-8746-a1e1e28c2907.png)

There is a downside to it, nan is good when you want to do something with the dataframe like any sort of mathematical operation because pandas can handle it internally. Wheres, '-' is visually more appealing but will make doing similar operations much more difficult for the user.

**Comment by stephenholleran** [comment]
> > > > @stephenholleran One more question in freq_table we replaced nan by 0.
> > > 
> > > 
> > > Eh, can't remember. Are you able to elaborate?
> > > > I have done the same for dist_matrix_by_dir_sector but this might not make sense when variable has below 0 values too.
> > > 
> > > 
> > > Sounds like it would be a problem if we do shear or temperature which can go negative.
> > 
> > 
> > Ah, this is to replace empty cells in the returned data and not something to do with redistributing wind speed nans. I have tried putting in '-'. It makes the table a lot more readable as you can quickly see the shape of the data. If we went with this we would just need to take them out when plotting?
> > ![image](https://user-images.githubusercontent.com/25622575/61869467-9a34e100-aed3-11e9-8746-a1e1e28c2907.png)
> 
> There is a downside to it, nan is good when you want to do something with the dataframe like any sort of mathematical operation because pandas can handle it internally. Wheres, '-' is visually more appealing but will make doing similar operations much more difficult for the user.

Why not just leave the NaN's?

**Comment by Inder95** [comment]
@ste

> * [x]  `freq_table()` is broken.
> 
> ![image](https://user-images.githubusercontent.com/25622575/61868817-2219eb80-aed2-11e9-9a1b-dc3f2d140181.png)

Fixed, somehow added an extra ',' at the end of ```plot_bins``` list which made python treat it as a tuple. 

![image](https://user-images.githubusercontent.com/23189856/62005398-8b943700-b12a-11e9-87ed-e4be33b8c2c3.png)


The only downside I can see to using the default bins of ``` ['0-3 m/s', '4-6 m/s', '7-9 m/s', '10-12 m/s', '13-15 m/s', '15+ m/s']``` is that when you plot other variables or bins like
```
tab = bw.freq_table(df.Spd40mN, df.Dir38mS, var_bin_array=[0, 8, 14, 41], var_bin_labels=['low', 'mid', 'high'],
                        direction_bin_array=[0, 90, 130, 200, 360],
                        direction_bin_labels=['northerly', 'easterly', 'southerly', 'westerly'],
                    plot_bins=[0, 8, 14, 41], plot_labels=None, return_data=True)
```
you need to specify plot_bins automatically because none will choose the 6 bins defined above.

**Comment by Inder95** [comment]
@stephenholleran Worked on most of the suggestions discussed in this thread. Feel free to merge.

**Comment by stephenholleran** [comment]
- [ ] The label for the rows, x-axis, is 'shear' when I would expect it to be 'Spd80mN'?

![image](https://user-images.githubusercontent.com/25622575/62637043-9de45100-b932-11e9-87ea-eebcf43d2c93.png)

Its a great plot straight out of the box.

**Comment by stephenholleran** [comment]
- [ ] I have trouble remembering which argument is for what. Can we rename `var_to_bin_series` to `var_to_bin_by_series` or something that is more descriptive. As it is I don't know if 'var_to_bin' is the variable I want to bin or the variable I want to bin by or against. Update the `var_bin_array` and `var_bin_labels` to be consistent.

**Comment by stephenholleran** [comment]
> The only downside I can see to using the default bins of ` ['0-3 m/s', '4-6 m/s', '7-9 m/s', '10-12 m/s', '13-15 m/s', '15+ m/s']` is that when you plot other variables or bins like
> 
> ```
> tab = bw.freq_table(df.Spd40mN, df.Dir38mS, var_bin_array=[0, 8, 14, 41], var_bin_labels=['low', 'mid', 'high'],
>                         direction_bin_array=[0, 90, 130, 200, 360],
>                         direction_bin_labels=['northerly', 'easterly', 'southerly', 'westerly'],
>                     plot_bins=[0, 8, 14, 41], plot_labels=None, return_data=True)
> ```
> 
> you need to specify plot_bins because none will choose the 6 bins defined above.

I think for `freq_table()` that is fine as it will mostly be used for wind speed. If someone wants to plot something else then they have the flexibility to change it or just use `dist_matrix_by_dir_sector()`.

**Comment by stephenholleran** [comment]
Is the freq_table wind rose not just a polar heat map?

**Comment by Inder95** [comment]
> Is the freq_table wind rose not just a polar heat map?

Not exactly, in wind_rose plot different colors can be stacked above one another in one sector because the same thing is plotted on that axis. Wheres in a polar heatmap these colors will have spaces between where the values of the 3rd variable can be empty.

It is tricky to explain but another way of looking at would be that would be the radial axis in polar plot is speed (or other variable here). What we is just draw a bar there and divide that bar according to the proportion of wind speeds in the wind_speed_bins but now there is another variable there with its own axis. Instead of drawing a bar on the wind speed axis we pick up each and every bin and color it according to the variable there. I think below is an example of how it would look like:
https://images.app.goo.gl/DZ8YJdTQ3XHZ6pNZ7

Also regarding the 3-D plot I found an example today just throwing it out here:
https://images.app.goo.gl/ZZSTWSyW7rbbkKEB7

**Comment by stephenholleran** [comment]
> > Is the freq_table wind rose not just a polar heat map?
> 
> Not exactly, in wind_rose plot different colors can be stacked above one another in one sector because the same thing is plotted on that axis. Wheres in a polar heatmap these colors will have spaces between where the values of the 3rd variable can be empty.
> 
> It is tricky to explain but another way of looking at would be that would be the radial axis in polar plot is speed (or other variable here). What we is just draw a bar there and divide that bar according to the proportion of wind speeds in the wind_speed_bins but now there is another variable there with its own axis. Instead of drawing a bar on the wind speed axis we pick up each and every bin and color it according to the variable there. 

ok, thanks.

> I think below is an example of how it would look like:
> https://images.app.goo.gl/DZ8YJdTQ3XHZ6pNZ7

Can matplotlib do this?
![image](https://user-images.githubusercontent.com/25622575/62690431-9b323c00-b9c4-11e9-8671-f52f5d4cb389.png)
 
> Also regarding the 3-D plot I found an example today just throwing it out here:
> https://images.app.goo.gl/ZZSTWSyW7rbbkKEB7

That looks great. Would we need to be able to let the user rotate it? How would that work in a pdf report?

**Comment by Inder95** [comment]
Will have to try that for matplotlib.

The 3-D plots can be rotated given that you use Python's plotting window rather than jupyter's inline method. For pdf report you will have to fix a view, one can play around in the interactive mode and then fix a view then download the graph as a 2-D image.

Depending on v1 release date we could skip this for now and come back to it later. But if we have time we can look into this.

**Comment by stephenholleran** [comment]
> Will have to try that for matplotlib.

Investigate this and if doable we could put it in v1.

> The 3-D plots can be rotated given that you use Python's plotting window rather than jupyter's inline method. For pdf report you will have to fix a view, one can play around in the interactive mode and then fix a view then download the graph as a 2-D image.
> 
> Depending on v1 release date we could skip this for now and come back to it later. But if we have time we can look into this.

Create an issue to remind us to come back to this? Cheers.

**Comment by stephenholleran** [comment]
- [ ] change colormap

Instead of using `YlGn` for the cmap use this instead:

```python
from matplotlib.colors import LinearSegmentedColormap
```

add to `bw_colors()`
```python
    elif bw_color == 'bw_col_map':
        colors = [(31 / 256, 41 / 256, 10 / 256),
                  (154 / 256, 205 / 256, 50 / 256),
                  (245 / 256, 250 / 256, 234 / 256)]
        bw_col_map = LinearSegmentedColormap.from_list('bw_col_map', colors, N=256)
        return bw_col_map
```

in `plot_dist_matrix()`
```python
    cm = ax.pcolormesh(matrix, cmap=bw_colors('bw_col_map'))
    # cm = ax.pcolormesh(matrix, cmap='YlGn')
```

This is close to the wind rose gradient but goes to a slightly darker green to make it more distinguishable.

![image](https://user-images.githubusercontent.com/25622575/62716899-cf285400-b9fa-11e9-95d6-e0b8218a48ff.png)

This should be useable in 12x24 contour plots and may even be able to replace the hard coded colors in `plot_rose_with_gradient()` too.

**for `plot_12x24_contours()`**
```python
x = ax.contourf(tab_12x24.columns, tab_12x24.index, tab_12x24.values, levels=levels,
        cmap=bw_colors('bw_col_map'))
        # colors=['#e1f0c1', '#d6ebad', '#c2e184', '#aed75b', '#9acd32', '#8ab92d', '#7ba428', '#6b9023'])
```
For updating the 12x24 this can be added to issue #118. You should also be able to increase the number of 'levels' now too.

**Comment by stephenholleran** [comment]
- [ ] Can you add label to the directional sector axis. e.g. Dir78mN.

  For me nothing comes up as you can see above and for Andy 'None' comes up. No idea why. It is best anyway to set the label to whatever direction series is used anyway so another reviewer will know what was used.

**Comment by Inder95** [comment]
![image](https://user-images.githubusercontent.com/23189856/63060382-efb84880-bee9-11e9-9176-bc0c77026a4c.png)

**Comment by Inder95** [comment]
@stephenholleran I think this branch is ready to be merged.

**Comment by stephenholleran** [comment]
> @stephenholleran I think this branch is ready to be merged.

What about the polar plot investigation? Or do you plan to try that in a different branch?
https://github.com/brightwind-dev/brightwind/issues/99#issuecomment-519446198

**Comment by Inder95** [comment]
> > @stephenholleran I think this branch is ready to be merged.
> 
> What about the polar plot investigation? Or do you plan to try that in a different branch?
> [#99 (comment)](https://github.com/brightwind-dev/brightwind/issues/99#issuecomment-519446198)

A different branch would be better here because we are not sure we will get there in time.

**Comment by stephenholleran** [comment]
- [ ] the var_to_bin label is taken from var_series.

**Comment by stephenholleran** [comment]
I see you removed the bw.mplstyle. As well as the Gotham font you want to take out it also changes some of the default font sizes and colours. See below the axis labels are tiny compared to the previous one.

Now:
![image](https://user-images.githubusercontent.com/25622575/63260364-a7e24a00-c278-11e9-870c-7f67e314453b.png)

Previous:
![image](https://user-images.githubusercontent.com/25622575/62716899-cf285400-b9fa-11e9-95d6-e0b8218a48ff.png)

This bw.mplstyle seems to set global formats even outside of the brightwind library. For coding this is bad practice to set global variables as the user may want to do something else but then will have to go back and change the global which will then have knock on consequences. Is there a way to set some standard formatting that is just limited to the library?

This should probably be put into a separate issue?

**Comment by Inder95** [comment]
> I see you removed the bw.mplstyle. As well as the Gotham font you want to take out it also changes some of the default font sizes and colours. See below the axis labels are tiny compared to the previous one.
> 
> Now:
> ![image](https://user-images.githubusercontent.com/25622575/63260364-a7e24a00-c278-11e9-870c-7f67e314453b.png)
> 
> Previous:
> ![image](https://user-images.githubusercontent.com/25622575/62716899-cf285400-b9fa-11e9-95d6-e0b8218a48ff.png)
> 
> This bw.mplstyle seems to set global formats even outside of the brightwind library. For coding this is bad practice to set global variables as the user may want to do something else but then will have to go back and change the global which will then have knock on consequences. Is there a way to set some standard formatting that is just limited to the library?
> 
> This should probably be put into a separate issue?

I commented it accidentally I guess when I was commenting out the font type used.. What do you mean by globally here, if you use matplotlib outside the library the plots don't have the parameters.

**Comment by stephenholleran** [comment]
> I commented it accidentally I guess when I was commenting out the font type used.. What do you mean by globally here, if you use matplotlib outside the library the plots don't have the parameters.

Neil has been using it in code to automate something and the mplstyle sets global parameters (like border color) that he doesn't want used. This isn't good behaviour.

**Comment by Inder95** [comment]
Strange because the documentation mentions that it does not change permanently (https://matplotlib.org/users/customizing.html). Anything you plot from the library will have those settings but matplotlib itself won't. If you import matplotlib elsewhere it shouldn't use bw.mplstyle. Can Neil describe exactly how is he using it?

**Comment by stephenholleran** [comment]
> Strange because the documentation mentions that it does not change permanently (https://matplotlib.org/users/customizing.html). Anything you plot from the library will have those settings but matplotlib itself won't. If you import matplotlib elsewhere it shouldn't use bw.mplstyle. Can Neil describe exactly how is he using it?

Lets focus on the other things to get this distribution finished. Just comment out the finding the font and create a new issue for this. The other comments and changes that I made are more important and the polar plot.

**Comment by Inder95** [comment]
> > Strange because the documentation mentions that it does not change permanently (https://matplotlib.org/users/customizing.html). Anything you plot from the library will have those settings but matplotlib itself won't. If you import matplotlib elsewhere it shouldn't use bw.mplstyle. Can Neil describe exactly how is he using it?
> 
> Lets focus on the other things to get this distribution finished. Just comment out the finding the font and create a new issue for this. The other comments and changes that I made are more important and the polar plot.

Already did that, you can merge this branch. I cannot put a number to the polar plot thing right now.

**Comment by stephenholleran** [comment]
What about:
https://github.com/brightwind-dev/brightwind/issues/99#issuecomment-522599627

And can you check out all the changes I made and make sure I didn't break anything? Renaming the dist_by_dir_sector func impacts on shear and and other things. Once all those ok you are free to merge.

**Comment by stephenholleran** [comment]
ok, getting an error with this.
![image](https://user-images.githubusercontent.com/25622575/63280951-aa0dce00-c2a3-11e9-89a4-311314097f94.png)

**Comment by Inder95** [comment]
> Edit:
> I'm on the new master branch.
> 
> How did this get past running the tests?

I had forgotten to run the tests when I made the changes to var_to_bin_series.name as you pointed out.

**Comment by AndyBrightWind** [comment]
Suggestion for both dist_matrix() and dist_matrix_by_dir_sector(): Include the aggregation method within the var_series label in the returned dataframe (similarly to the contour bar label in the plot). This will be clearer for end users of the dataframe who haven't seen how it was generated.

**Comment by stephenholleran** [comment]
> Suggestion for both dist_matrix() and dist_matrix_by_dir_sector(): Include the aggregation method within the var_series label in the returned dataframe (similarly to the contour bar label in the plot). This will be clearer for end users of the dataframe who haven't seen how it was generated.

Done.

@Inder95 can you run the tests please?

**Comment by Inder95** [comment]
> > Suggestion for both dist_matrix() and dist_matrix_by_dir_sector(): Include the aggregation method within the var_series label in the returned dataframe (similarly to the contour bar label in the plot). This will be clearer for end users of the dataframe who haven't seen how it was generated.
> 
> Done.
> 
> @Inder95 can you run the tests please?

Test failed, will have a look into this, the last line seems wrong
```

   if var_label is None:
        var_series.name = var_series.name + '_distributed'	        var_label = aggregation_method.capitalize() + ' of ' + var_series.name
    if var_series.name == x_series.name:	    var_series.name = var_label
```
Tests for plot_scatter also failed.

**Comment by Inder95** [comment]
All tests passed now. There was a wrong test for freq_table sitting in test_transform, fixed it. For plot_scatter the issue was passing series in the format df[[col_name]], fixed that too.
